File: depth-first-search/count-the-number-of-complete-components.java
import java.util.Dictionary;
import java.util.Hashtable;
import java.util.HashSet;
import java.util.Set;
import java.util.ArrayList;
class Solution {
    public int countCompleteComponents(int n, int[][] edges) {
        Dictionary<Integer, Set<Integer>> edges_dict = new Hashtable<>();
        for(int[] edge : edges){
            if(edges_dict.get(edge[0]) == null){
                edges_dict.put(edge[0], new HashSet<>());
            }
            edges_dict.get(edge[0]).add(edge[1]);

            if(edges_dict.get(edge[1]) == null){
                edges_dict.put(edge[1], new HashSet<>());
            }
            edges_dict.get(edge[1]).add(edge[0]);
        }

        List<Set<Integer>> components = new ArrayList<Set<Integer>>();
        int curr_index = 0;
        Dictionary<Integer, Integer> nodes_to_component = new Hashtable<>();
        for(int i = 0;i < n;i++){
            if(nodes_to_component.get(i) == null){
                Set<Integer> curr_component = new HashSet<>();
                curr_component.add(i);
                List<Integer> stack = new ArrayList<Integer>();
                stack.add(i);
                while(stack.size() > 0){
                    int curr_node = stack.remove(stack.size() - 1);
                    if(edges_dict.get(curr_node) != null){
                        for(int end_node : edges_dict.get(curr_node)){
                            if(!curr_component.contains(end_node)){
                                curr_component.add(end_node);
                                nodes_to_component.put(end_node, curr_index);
                                stack.add(end_node);
                            }
                        }
                    }
                }
                curr_index += 1;
                components.add(curr_component);
            }
        }
        
        boolean[] answer_list = new boolean[components.size()];
        for(int new_node = 0;new_node < n;new_node++){
            if(edges_dict.get(new_node) != null && nodes_to_component.get(new_node) != null && edges_dict.get(new_node).size() != components.get(nodes_to_component.get(new_node)).size() - 1){
                answer_list[nodes_to_component.get(new_node)] = true;
            }
        }

        int answer = 0;
        for(int j = 0;j < answer_list.length;j++){
            if(!answer_list[j]){
                answer += 1;
            }
        }
        return answer;
    }
}

File: union-find/count-the-number-of-complete-components.py
class Solution:
    def countCompleteComponents(self, n: int, edges: List[List[int]]) -> int:
        edges_dict = defaultdict(set)
        for start, end in edges:
            edges_dict[start].add(end)
            edges_dict[end].add(start)
        
        components = []
        curr_index = 0
        nodes_to_component = {}
        for i in range(n):
            if i not in nodes_to_component:
                curr_component = set()
                curr_component.add(i)
                queue = deque()
                queue.append(i)
                nodes_to_component[i] = curr_index
                while queue:
                    curr_node = queue.popleft()
                    for end_node in edges_dict[curr_node]:
                        if end_node not in curr_component:
                            curr_component.add(end_node)
                            nodes_to_component[end_node] = curr_index
                            queue.append(end_node)
                curr_index += 1
                components.append(curr_component)

        answer = [True] * len(components)
        for curr_node in edges_dict:
            if len(edges_dict[curr_node]) != len(components[nodes_to_component[curr_node]]) - 1:
                answer[nodes_to_component[curr_node]] = False
        return sum(answer)